boa_engine/src/builtins/generator/mod.rs

```diff
@@ -0,0 +1,412 @@
+        })?;
+        let mut generator_obj_mut = generator_obj.borrow_mut();
+        let generator = generator_obj_mut.as_generator_mut().ok_or_else(|| {
+            context.construct_type_error("generator resumed on non generator object")
+        })?;
+        let mut state = generator.state;
+
+        if state == GeneratorState::Executing {
+            return Err(context.construct_type_error("Generator should not be executing"));
+        }
+
+        // 2. If state is suspendedStart, then
+        if state == GeneratorState::SuspendedStart {
+            // a. Set generator.[[GeneratorState]] to completed.
+            generator.state = GeneratorState::Completed;
+            // b. Once a generator enters the completed state it never leaves it and its associated execution context is never resumed. Any execution state associated with generator can be discarded at this point.
+            generator.context = None;
+            // c. Set state to completed.
+            state = GeneratorState::Completed;
+        }
+
+        // 3. If state is completed, then
+        if state == GeneratorState::Completed {
+            // a. If abruptCompletion.[[Type]] is return, then
+            if let Ok(value) = abrupt_completion {
+                // i. Return CreateIterResultObject(abruptCompletion.[[Value]], true).
+                return Ok(create_iter_result_object(value, true, context));
+            }
+            // b. Return Completion(abruptCompletion).
+            return abrupt_completion;
+        }
+
+        // 4. Assert: state is suspendedYield.
+        // 5. Let genContext be generator.[[GeneratorContext]].
+        // 6. Let methodContext be the running execution context.
+        // 7. Suspend methodContext.
+        // 8. Set generator.[[GeneratorState]] to executing.
+        // 9. Push genContext onto the execution context stack; genContext is now the running execution context.
+        // 10. Resume the suspended evaluation of genContext using abruptCompletion as the result of the operation that suspended it. Let result be the completion record returned by the resumed computation.
+        // 11. Assert: When we return here, genContext has already been removed from the execution context stack and methodContext is the currently running execution context.
+        // 12. Return Completion(result).
+        let generator_context_cell = generator
+            .context
+            .take()
+            .expect("generator context cannot be empty here");
+        let mut generator_context = generator_context_cell.borrow_mut();
+
+        generator.state = GeneratorState::Executing;
+        drop(generator_obj_mut);
+
+        std::mem::swap(
+            &mut context.realm.environments,
+            &mut generator_context.environments,
+        );
+        std::mem::swap(&mut context.vm.stack, &mut generator_context.stack);
+        context.vm.push_frame(generator_context.call_frame.clone());
```

Comment:
Would having a proper execution context stack make it a bit easier to switch context, instead of having to manually swap them?